### PR TITLE
Some optimizations for constant conditions in JOIN ON

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add optimizations for constant conditions in JOIN ON, ref #26928

### CODE REVIEW

**Comment by vdimir** [comment]
Do not merge before https://github.com/ClickHouse/ClickHouse/pull/21320 because of conflicts

**Review by KochetovNicolai — APPROVED** [review]
Ok, just refactoring in general.

**Comment by alexey-milovidov** [comment]
@mergifyio update

File: src/Interpreters/MergeJoin.cpp [hunk 1]
```diff
@@ -512,6 +513,7 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
 
     const auto & onexpr = table_join->getOnlyClause();
     std::tie(mask_column_name_left, mask_column_name_right) = onexpr.condColumnNames();
+    LOG_DEBUG(log, "Right sample block: {}", right_sample_block.dumpStructure());
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/MergeJoin.cpp:515]:
```suggestion
```

**Pull request merged**